### PR TITLE
Add support for SPV_KHR_constant_data and SPV_KHR_abort

### PR DESCRIPTION
**Pull request by Samsung-CraigG**
Used by VK_KHR_shader_abort / VK_KHR_shader_constant_data

### CODE REVIEW

**Comment by ShchchowAMD** [comment]
Hi, I've created a new commit for submission under branch:
https://github.com/ShchchowAMD/glslang/tree/abortKHR

Please feel free to tell me what extra works needed if anything works not as expected, sorry for blocking PR submission.
Thanks!

File: SPIRV/GlslangToSpv.cpp [hunk 1]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+        int specifierIndex; // -1 if not a specifier.
+        constStrInfo(glslang::TString str, int spec) : string(str), specifierIndex(spec){};
+    } strInfo;
+
+    std::vector<strInfo> splitedStr, tempSplitedStr;
+    const uint32_t formatSpecifiersSize = 4;
+    const char* formatSpecifiers[formatSpecifiersSize] = {"%d", "%i", "%f", "%u"};
+    // 1. Split original message string with format specifiers.
+    const glslang::TString* msg = !isEmptyMsg ? glslangOperands[0]->getAsConstantUnion()->getConstArray()[0].getSConst()
+                                             : new glslang::TString("\0");
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3160]: `msg` will own the heap allocation if the message is empty. This will likely create leaks. Could you either allocate on the stack or restructure the branches so no heap allocation is needed?

File: SPIRV/GlslangToSpv.cpp [hunk 2]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+    builder.addExtension(spv::E_SPV_KHR_constant_data);
+    builder.addExtension(spv::E_SPV_KHR_abort);
+
+    typedef struct constStrInfo {
+        glslang::TString string;
+        int specifierIndex; // -1 if not a specifier.
+        constStrInfo(glslang::TString str, int spec) : string(str), specifierIndex(spec){};
+    } strInfo;
+
+    std::vector<strInfo> splitedStr, tempSplitedStr;
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3155]: grammar nit: s/`splitedStr`/`splitStr`/ and s/`tempSplitedStr`/`tempSplitStr`/

File: SPIRV/GlslangToSpv.cpp [hunk 3]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+            int specifierIndex = splitedStr[j].specifierIndex;
+            auto pos = str.find(formatSpecifiers[i]);
+            while (pos != std::string::npos) {
+                tempSplitedStr.push_back(strInfo(str.substr(0, pos), specifierIndex));
+                tempSplitedStr.push_back(strInfo(glslang::TString(formatSpecifiers[i]), i));
+                str = str.substr(pos + strlen(formatSpecifiers[i]));
+                pos = str.find(formatSpecifiers[i]);
+            }
+            if (str.size() > 0 || isEmptyMsg)
+                tempSplitedStr.push_back(strInfo(str, specifierIndex));
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3174]: nit: Check for `!str.empty()` instead of `str.size() > 0`.

Note that `isEmptyMsg` is set when `glslangOperands` is empty (no arguments). If the caller passes an empty string (e.g., `abortEXT("", v);`) then `msg` will have size 0. This will make the loop not push anything into `tempSplitedStr`. When it goes on to pop from `structMemberOffsets`, it will leave the vector empty. So when the code tries to access `structMemberOffsets.back()` the vector is empty (undefined behaviour).

This needs a test for an empty string and special handling for empty string literals (`""`).

File: glslang/MachineIndependent/Versions.cpp [hunk 4]
```diff
@@ -553,6 +555,7 @@ void TParseVersions::getPreamble(std::string& preamble)
             "#define GL_EXT_shared_memory_block 1\n"
             "#define GL_EXT_shader_integer_mix 1\n"
             "#define GL_EXT_spec_constant_composites 1\n"
+            "#define E_GL_EXT_abort 1\n"
```
Comment (dnovillo) [inline glslang/MachineIndependent/Versions.cpp:558]: s/`E_GL_EXT_abort`/`GL_EXT_abort`/ since `E_GL_EXT_abort` is the C++ constant in `Versions.h`. The GLSL injection should use the GLSL extension name.

File: SPIRV/SpvBuilder.cpp [hunk 5]
```diff
@@ -3380,6 +3383,12 @@ void Builder::createNoResultOp(Op opCode, const std::vector<Id>& operands)
     addInstruction(std::unique_ptr<Instruction>(op));
 }
 
+void Builder::makeAbortKHR(const std::vector<Id>& operands)
```
Comment (dnovillo) [inline SPIRV/SpvBuilder.cpp]: `Builder` already has a two-argument overload of `makeStatementTerminator` for block terminators that take operands. It's identical to the new `makeAbortKHR`. Could you remove `makeAbortKHR` and call something like `builder.makeStatementTerminator(spv::Op::OpAbortKHR, {structLoadType, messageVar}, "post-abort")` directly?

File: SPIRV/SpvBuilder.cpp [hunk 6]
```diff
@@ -3468,6 +3477,29 @@ Id Builder::createTriOp(Op opCode, Id typeId, Id op1, Id op2, Id op3)
...
+    for (auto id : operands)
+        op->addStringOperand(id);
+    module.mapInstruction(op);
+    constantsTypesGlobals.push_back(std::unique_ptr<Instruction>(op));
+
+    return op->getResultId();
+}
+
+Id Builder::createSpecConst(Op opCode, Id typeId, const unsigned int literalOp)
+{
```
Comment (dnovillo) [inline SPIRV/SpvBuilder.cpp:3487]: This is equivalent to builder.makeUintConstant(strElemLen, /*specConstant=*/true). Could you remove `createSpecConst` and use the existing overload?

File: SPIRV/GlslangToSpv.cpp [hunk 7]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+    builder.addExtension(spv::E_SPV_KHR_abort);
+
+    typedef struct constStrInfo {
+        glslang::TString string;
+        int specifierIndex; // -1 if not a specifier.
+        constStrInfo(glslang::TString str, int spec) : string(str), specifierIndex(spec){};
+    } strInfo;
+
+    std::vector<strInfo> splitedStr, tempSplitedStr;
+    const uint32_t formatSpecifiersSize = 4;
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3156]: I don't recall whether this string splitting is required by the spec or whether the spec allows a simpler representation like a single `OpConstantDataKHR` for the entire string. If it doesn't, maybe this could use a simpler lowering like we do in `debugPrintfEXT`. Can you confirm whether the spec mandates this and also what are the rules about padding?

File: SPIRV/SpvBuilder.cpp [hunk 8]
```diff
@@ -3468,6 +3477,29 @@ Id Builder::createTriOp(Op opCode, Id typeId, Id op1, Id op2, Id op3)
     return op->getResultId();
 }
 
+Id Builder::createConstData(Op opCode, Id typeId, const std::vector<const char*> operands)
+{
+    Instruction* op = new Instruction(getUniqueId(), typeId, opCode);
+    op->reserveOperands(operands.size());
+    for (auto id : operands)
+        op->addStringOperand(id);
```
Comment (dnovillo) [inline SPIRV/SpvBuilder.cpp:3479]: Does the encoding use `LiteralString` or `LiteralBytes`?

File: SPIRV/GlslangToSpv.cpp [hunk 9]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+void TGlslangToSpvTraverser::createAbortEXT(const glslang::TIntermSequence glslangOperands)
+{
+    bool isEmptyMsg = glslangOperands.size() == 0;
+    // Add Capability and extensions.
+    builder.addCapability(spv::Capability::AbortKHR);
+    builder.addCapability(spv::Capability::ConstantDataKHR);
+    builder.addExtension(spv::E_SPV_KHR_constant_data);
+    builder.addExtension(spv::E_SPV_KHR_abort);
+
+    typedef struct constStrInfo {
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3149]: nit. No `typedef` necessary. This can use `strInfo` or `constStrInfo`. No need to have both.

File: SPIRV/GlslangToSpv.cpp [hunk 10]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
...
+        constStrInfo(glslang::TString str, int spec) : string(str), specifierIndex(spec){};
+    } strInfo;
+
+    std::vector<strInfo> splitedStr, tempSplitedStr;
+    const uint32_t formatSpecifiersSize = 4;
+    const char* formatSpecifiers[formatSpecifiersSize] = {"%d", "%i", "%f", "%u"};
+    // 1. Split original message string with format specifiers.
+    const glslang::TString* msg = !isEmptyMsg ? glslangOperands[0]->getAsConstantUnion()->getConstArray()[0].getSConst()
+                                             : new glslang::TString("\0");
+    splitedStr.push_back(strInfo(*const_cast<glslang::TString*>(msg), -1));
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3161]: No need to `const_cast` here. The constructor takes `str` by value. This can be `strInfo(*msg, -1)`.

File: SPIRV/GlslangToSpv.cpp [hunk 11]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
     return builder.createCompositeConstruct(resultTypeId, constituents);
 }
 
+void TGlslangToSpvTraverser::createAbortEXT(const glslang::TIntermSequence glslangOperands)
+{
+    bool isEmptyMsg = glslangOperands.size() == 0;
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3142]: s/`.size() == 0`/`.empty()`/

File: SPIRV/GlslangToSpv.cpp [hunk 12]
```diff
@@ -3136,6 +3137,101 @@ spv::Id TGlslangToSpvTraverser::createCompositeConstruct(spv::Id resultTypeId, s
     return builder.createCompositeConstruct(resultTypeId, constituents);
 }
 
+void TGlslangToSpvTraverser::createAbortEXT(const glslang::TIntermSequence glslangOperands)
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp:3140]: this will copy the vector on every call. Pass it by const reference, please.

**Review by dnovillo — CHANGES_REQUESTED** [review]

**Comment by Samsung-CraigG** [comment]
Hi Chow – couple of commits I made that will need applying to your new branch:

 

Build fix for GCC:

https://github.com/KhronosGroup/glslang/pull/4196/changes/2140ff598e8858d478f0440ea09669887b5571f4#diff-4c2971e76ad3cea3aee5bc8b1496244b0f2e891a4fabed3582990129c938925c

(otherwise it’ll fail CI)

 

OpAbortKHR should be the last instruction in a block:

https://github.com/KhronosGroup/glslang/pull/4196/changes/71eef953a730356c32a19e3f56601c7650aef63e

Remove Builder::makeAbortKHR():

https://github.com/KhronosGroup/glslang/pull/4196/changes/d21961039432ffa5a900cf9755ab75717df7c44b

 

To enforce the spec rule that states “This instruction must be the last instruction in a block.”.

 

Regards,

Craig.

**Comment by ShchchowAMD** [comment]
> Hi Chow – couple of commits I made that will need applying to your new branch: Build fix for GCC: [2140ff5#diff-4c2971e76ad3cea3aee5bc8b1496244b0f2e891a4fabed3582990129c938925c](https://github.com/KhronosGroup/glslang/pull/4196/changes/2140ff598e8858d478f0440ea09669887b5571f4#diff-4c2971e76ad3cea3aee5bc8b1496244b0f2e891a4fabed3582990129c938925c) (otherwise it’ll fail CI) OpAbortKHR should be the last instruction in a block: [71eef95](https://github.com/KhronosGroup/glslang/pull/4196/changes/71eef953a730356c32a19e3f56601c7650aef63e) Remove Builder::makeAbortKHR(): [d219610](https://github.com/KhronosGroup/glslang/pull/4196/changes/d21961039432ffa5a900cf9755ab75717df7c44b) To enforce the spec rule that states “This instruction must be the last instruction in a block.”. Regards, Craig. From: Chow ***@***.***> Sent: 05 April 2026 11:27 To: KhronosGroup/glslang ***@***.***> Cc: Craig Graham ***@***.***>; Mention ***@***.***> Subject: Re: [KhronosGroup/glslang] Add support for SPV_KHR_constant_data and SPV_KHR_abort (PR #4196) ShchchowAMD left a comment ([KhronosGroup/glslang#4196](https://github.com/KhronosGroup/glslang/pull/4196)) <https://protect2.fireeye.com/v1/url?k=8d453406-ecce2123-8d44bf49-74fe485cbff6-325074c4d3a45289&q=1&e=55899f22-338a-4cf3-b55b-8a7432e521c2&u=https%3A%2F%2Fgithub.com%2FKhronosGroup%2Fglslang%2Fpull%2F4196%23issuecomment-4188667060> Hi, I've created a new commit for submission under branch: https://github.com/ShchchowAMD/glslang/tree/abortKHR <https://protect2.fireeye.com/v1/url?k=113480d0-70bf95f5-11350b9f-74fe485cbff6-00461b2f5176b82c&q=1&e=55899f22-338a-4cf3-b55b-8a7432e521c2&u=https%3A%2F%2Fgithub.com%2FShchchowAMD%2Fglslang%2Ftree%2FabortKHR> Please feel free to tell me what extra works needed if anything works not as expected, sorry for blocking PR submission. Thanks! — Reply to this email directly, <https://protect2.fireeye.com/v1/url?k=4af55c31-2b7e4914-4af4d77e-74fe485cbff6-e536a3a1cc376389&q=1&e=55899f22-338a-4cf3-b55b-8a7432e521c2&u=https%3A%2F%2Fgithub.com%2FKhronosGroup%2Fglslang%2Fpull%2F4196%23issuecomment-4188667060> view it on GitHub, or <https://protect2.fireeye.com/v1/url?k=3e283c4b-5fa3296e-3e29b704-74fe485cbff6-a8ab8bc6bc2b8bbc&q=1&e=55899f22-338a-4cf3-b55b-8a7432e521c2&u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FB7GMNEH57BOONUQ7NVUZAXD4UIYF5AVCNFSM6AAAAACW4I6IKGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DCOBYGY3DOMBWGA> unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>

Hi, Sure.

Sorry for the late reply.
Please feel free to push commits into that branch or let me know if I need to do this work. I'll make commits to fix codes according to above reviews this week.

File: SPIRV/SpvBuilder.cpp (hunk 8 above)
Comment (ShchchowAMD) [reply SPIRV/SpvBuilder.cpp:3479]: It is 8 bit width elements here, so should be Bytes I think?
https://gitlab.khronos.org/spirv/spirv-extensions/-/blob/b75fc40c686119e6590d1eabfabe2f843dc30e3a/SPV_KHR_constant_data.asciidoc

File: SPIRV/GlslangToSpv.cpp (hunk 7 above)
Comment (ShchchowAMD) [reply SPIRV/GlslangToSpv.cpp:3156]: https://gitlab.khronos.org/GLSL/GLSL/-/merge_requests/132/diffs

Spec mentioned in the end as:

    The set of format specifiers is implementation-dependent, but must
    include at least "%d" and "%i" (int), "%u" (uint), and "%f" (float).

Comment (ShchchowAMD) [reply SPIRV/GlslangToSpv.cpp:3156]: If I remember correctly, data is not simple char type now, but sub-strings. Discussed internally before.
https://gitlab.khronos.org/GLSL/glslang/-/merge_requests/125

File: SPIRV/GlslangToSpv.cpp (hunk 3 above)
Comment (ShchchowAMD) [reply SPIRV/GlslangToSpv.cpp:3174]: Yea. I'm now treating it as an empty string.

**Comment by Samsung-CraigG** [comment]
Closed as superseded by #4222
